File: scripts/test_phase6j_concurrency.js
const axios = require('axios');
const crypto = require('crypto');

const BASE_URL = 'http://localhost:3000/api';

async function main() {
    console.log("Starting Concurrency Tests for Domains");

    // 1. Create Admin and Workspace
    const r1 = await axios.post(`${BASE_URL}/auth/register`, {
        name: `Admin Concur ${Date.now()}`,
        email: `admin_${Date.now()}@test.com`,
        password: 'password123'
    });
    const token = r1.data.token;

    const r2 = await axios.post(`${BASE_URL}/workspaces`, { name: 'WS Concur' }, { headers: { Authorization: `Bearer ${token}` } });
    const wsId = r2.data.workspace.id;

    // 2. Create Application A and Application B
    const r3 = await axios.post(`${BASE_URL}/workspaces/${wsId}/applications`, { name: 'App A', slug: `appa-${Date.now()}` }, { headers: { Authorization: `Bearer ${token}` } });
    const appA = r3.data.application;

    const r4 = await axios.post(`${BASE_URL}/workspaces/${wsId}/applications`, { name: 'App B', slug: `appb-${Date.now()}` }, { headers: { Authorization: `Bearer ${token}` } });
    const appB = r4.data.application;

    // 3. Request Access Tokens
    const rA = await axios.post(`${BASE_URL}/applications/${appA.id}/auth/request-access`, {}, { headers: { Authorization: `Bearer ${token}` } });
    const tokenA = rA.data.token;

    const rB = await axios.post(`${BASE_URL}/applications/${appB.id}/auth/request-access`, {}, { headers: { Authorization: `Bearer ${token}` } });
    const tokenB = rB.data.token;

    const testHostname = `race-${Date.now()}.com`;

    console.log(`[ ] Race creating domain ${testHostname} on two apps simultaneously...`);

    const p1 = axios.post(`${BASE_URL}/applications/${appA.id}/domains`, { hostname: testHostname, type: 'CUSTOM_DOMAIN' }, { headers: { Authorization: `Bearer ${tokenA}` } });
    const p2 = axios.post(`${BASE_URL}/applications/${appB.id}/domains`, { hostname: testHostname, type: 'CUSTOM_DOMAIN' }, { headers: { Authorization: `Bearer ${tokenB}` } });

    let successCount = 0;
    let failCount = 0;

    const results = await Promise.allSettled([p1, p2]);
    results.forEach(res => {
        if (res.status === 'fulfilled') successCount++;
        else failCount++;
    });

    console.log(`[+] Successes: ${successCount}, Failures: ${failCount}`);
    if (successCount !== 1 || failCount !== 1) {
        console.error("[!] Concurrency test failed. One and only one should succeed.");
        process.exit(1);
    }
    
    console.log("[+] Concurrency test passed.");
}

main().catch(console.error);

File: scripts/test_phase6j_concurrency.py
import requests
import threading
import uuid
import time
import os

BASE_URL = os.getenv('API_URL', 'http://localhost:3000/api')
results = []

def register_admin():
    email = f"admin_{uuid.uuid4()}@test.com"
    pw = "password123"
    requests.post(f"{BASE_URL}/auth/register", json={"name": "Admin", "email": email, "password": pw})
    res = requests.post(f"{BASE_URL}/auth/login", json={"email": email, "password": pw})
    return res.json().get('token')

def create_domain(app_id, token, hostname):
    res = requests.post(f"{BASE_URL}/applications/{app_id}/domains", json={
        "hostname": hostname, "type": "CUSTOM_DOMAIN"
    }, headers={"Authorization": f"Bearer {token}"})
    results.append(res.status_code)

def main():
    token = register_admin()
    headers = {"Authorization": f"Bearer {token}"}
    
    ws_res = requests.post(f"{BASE_URL}/workspaces", json={"name": "WS"}, headers=headers)
    ws_id = ws_res.json()['workspace']['id']

    appA_res = requests.post(f"{BASE_URL}/workspaces/{ws_id}/applications", json={"name": "App A", "slug": f"a{uuid.uuid4().hex[:5]}", "authentication_enabled": True}, headers=headers)
    appB_res = requests.post(f"{BASE_URL}/workspaces/{ws_id}/applications", json={"name": "App B", "slug": f"b{uuid.uuid4().hex[:5]}", "authentication_enabled": True}, headers=headers)
    
    appA_id = appA_res.json()['application']['id']
    appB_id = appB_res.json()['application']['id']

    tokenA = requests.post(f"{BASE_URL}/applications/{appA_id}/auth/login", json={"email": "a@test.com", "password": "123"}).json() # Will fail
    tokenA = requests.post(f"{BASE_URL}/applications/{appA_id}/auth/request-access", headers=headers).json() # admin is member
    
    # Wait, the admin created the apps, so they just login to app
    resA = requests.post(f"{BASE_URL}/applications/{appA_id}/auth/request-access", headers=headers) # we just use platform token? no we can't
    # Wait, platform token works for platform endpoints, application endpoints need app token if using /api/applications/...? 
    # NO! Admin domains endpoints use platform token since they are `/api/applications/:id/domains` !!
    
    test_host = f"race-{uuid.uuid4().hex[:6]}.com"
    print(f"[*] Racing {test_host} on two apps...")
    
    t1 = threading.Thread(target=create_domain, args=(appA_id, token, test_host))
    t2 = threading.Thread(target=create_domain, args=(appB_id, token, test_host))
    
    t1.start()
    t2.start()
    
    t1.join()
    t2.join()
    
    success = results.count(201)
    fails = len(results) - success
    
    print(f"[+] Successes: {success}, Failures: {fails}")
    if success == 1 and fails == 1:
        print("[+] Concurrency Race Check PASSED")
    else:
        print("[!] Concurrency Race Check FAILED")
        exit(1)

if __name__ == "__main__":
    main()
